src/lib/services/group-service.ts

```diff
@@ -103,43 +108,47 @@ export class GroupService {
 
         await this.validateGroup(group, preData);
 
-        const newGroup = await this.groupStore.update(group);
+        return this.db.transaction(async (tx) => {
```

Comment:
Do we really want to expose entire knex to all services? Or could transactional stores provide a transaction function (possible via a mixing) instead?

Reply:
Maybe we could offer a special u`withTransaction` function in the `stores` container you can use instead of exposing db/knex directly to all services.

Reply:
There's a reason, maybe a bad one but a reason - it makes a lot more sense in terms of how you end up using this. If we're moving transaction responsibilities into the service layers it makes more sense for the service to both own and start the transaction. You can reach into one of the stores to make that happen but you end up with things like this:

```js
const someStore = getStore();
const someOtherStore = getStore();

someStore.transaction(async (tx) => { //why is this store starting the transaction? Could it be the other one? That's not clear from the syntax 
})

``` 
That's probably not worth the tradeoff of exposing the entire Knex object to the services but we could pass a closure that yields a transaction object. Something like

```
const transactionMaker(knex: Knex){
   startTransaction(){
      return knex.transction()
   }
   return startTransction
}

```
That would allow us to only provide access to creating a transaction to the services while still maintaining a nice API. 

WDYT?

Reply:
yes I kind of corrected myself in the second comment. Transactions should be able to span stores, that is a important need. I am not to worried what we call it, but allowing the services to start a transaction without interacting directly with the database is important to me. 

I am thinking that we could envision a possible future where the entire store/database layer would be replaced with an in-memory implementation at one point (to facilitate a throw-away unleash you spin up for automated integration testing), then it would be nice to not tie the transaction handling directly to knex, but instead provide an abstraction we can replicate with other store adapters as well.

Reply:
Right, so I've done some stuff here. The services can now depend on a `TransactionCreator<UnleashTransaction>` type instead of directly on anything related to Knex. `TransactionCreator<T>` is basically an alias for a closure type where the closed function is generic over an inner type that's determined by how it's called - that's to allow you to have a start a transaction that will resolve to an `IGroup` for example (S would be `IGroup` in that case). The `UnleashTransaction` is an alias for `KnexTransaction | null`, that doesn't do anything special at the moment, it's just an escape hatch so that  you can mock out the transaction handling in the services and not have to know about or use Knex. So the service call now looks like this:

```js
export class GroupService {
    private startTransaction: TransactionCreator<UnleashTransaction>;

    constructor(
        stores: Pick<IUnleashStores, 'groupStore' | 'eventStore' | 'userStore'>,
        { getLogger }: Pick<IUnleashConfig, 'getLogger'>,
        startTransaction: TransactionCreator<UnleashTransaction>,
    ) {
        //some fields we're ignoring
        this.startTransaction = startTransaction;
    }
   
    //totally made up method
    async getGroups(): Promise<IGroup> {
    
        return this.startTransaction(async (tx) => {  // <- the new magic
            return await this.groupStore
                .transactional(tx)
                .getAllGroups();
        });
    }
}
```

I've also made a little utulity method to generate an `TransactionCreator<KnexTransaction>` to actually use this. That's as simple as doing:

```js
    const db: Knex = someCallThatReturnsAKnex(); 
    const startTransaction = createKnexTransactionStarter(db);
```

I've also cleaned up the FakeGroupStore and used that in a test to show that this can actually be mocked out relatively easily.

If we want to push this in the future, I think we could do something like

```js
class GroupService<T> {
  
  constructor(store: IGroupStore<T>) {
    //stuff
  }
}
```
So that the service transaction type can be type bound to the type of the stores being passed in but I haven't explored that in detail, this is already a fairly big change and that would mean mucking with the types on the stores.

@ivarconr WDYT?